clients/rospy/src/rospy/core.py

```diff
@@ -646,4 +648,20 @@ def xmlrpcapi(uri):
     uriValidate = urlparse.urlparse(uri)
     if not uriValidate[0] or not uriValidate[1]:
         return None
-    return xmlrpcclient.ServerProxy(uri)
+    if not cache:
+        return xmlrpcclient.ServerProxy(uri)
+    if uri not in _xmlrpc_cache:
+        _xmlrpc_cache[uri] = _LockedServerProxy(uri)
+    return _xmlrpc_cache[uri]
+
+
+class _LockedServerProxy(xmlrpcclient.ServerProxy):
+
+    def __init__(self, *args, **kwargs):
+        xmlrpcclient.ServerProxy.__init__(self, *args, **kwargs)
+        self._lock = threading.Lock()
+
+    def _ServerProxy__request(self, methodname, params):
```

Comment:
Why are you not using ``__request`` directly here?

Reply:
This is unfortunately the way to override private methods in python.

Reply:
Ah, ok. (explanation can be found [here](https://docs.python.org/2/tutorial/classes.html#private-variables-and-class-local-references))